### PR TITLE
Use Java collections in GenericDataFile to fix Kryo serialization

### LINKED ISSUE #446: KryoException when writing Iceberg tables in Spark
Say we have a table `test (col1: string)`.
```
case class Input(col1: String)
val df = spark.createDataFrame(List(Input("data1"), Input("data2")))
df.write.format("iceberg").mode("overwrite").option("write-format", "parquet").save("default.test")
```
```
2019-09-04 05:35:13 ERROR TaskResultGetter:91 - Exception while getting task result
com.esotericsoftware.kryo.KryoException: java.lang.UnsupportedOperationException
Serialization trace:
splitOffsets (org.apache.iceberg.GenericDataFile)
files (org.apache.iceberg.spark.source.Writer$TaskCommit)
	at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:144)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:543)
	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:731)
	at com.esotericsoftware.kryo.serializers.DefaultArraySerializers$ObjectArraySerializer.read(DefaultArraySerializers.java:391)
	at com.esotericsoftware.kryo.serializers.DefaultArraySerializers$ObjectArraySerializer.read(DefaultArraySerializers.java:302)
	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:731)
	at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:543)
	at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:813)
	at org.apache.spark.serializer.KryoSerializerInstance.deserialize(KryoSerializer.scala:362)
	at org.apache.spark.scheduler.DirectTaskResult.value(TaskResult.scala:88)
	at org.apache.spark.scheduler.TaskResultGetter$$anon$3$$anonfun$run$1.apply$mcV$sp(TaskResultGetter.scala:72)
	at org.apache.spark.scheduler.TaskResultGetter$$anon$3$$anonfun$run$1.apply(TaskResultGetter.scala:63)
	at org.apache.spark.scheduler.TaskResultGetter$$anon$3$$anonfun$run$1.apply(TaskResultGetter.scala:63)
	at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1945)
	at org.apache.spark.scheduler.TaskResultGetter$$anon$3.run(TaskResultGetter.scala:62)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.UnsupportedOperationException
	at org.apache.iceberg.shaded.com.google.common.collect.ImmutableCollection.add(ImmutableCollection.java:244)
	at com.esotericsoftware.kryo.serializers.CollectionSerializer.read(CollectionSerializer.java:134)
	at com.esotericsoftware.kryo.serializers.CollectionSerializer.read(CollectionSerializer.java:40)
	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:731)
	at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
	... 18 more
2019-09-04 05:35:13 ERROR WriteToDataSourceV2Exec:70 - Data source writer IcebergWrite(table=default.test, format=PARQUET) is aborting.
2019-09-04 05:35:13 ERROR WriteToDataSourceV2Exec:70 - Data source writer IcebergWrite(table=default.test, format=PARQUET) aborted.
```

Hitting this when writing Parquet data in Spark 2.4 using the latest master. I believe this is because in 549da809490976b53a13b14596dd240ed74bce5e an immutable copy of the split offsets are stored in GenericDataFile. The serialization is successful, but when deserializing, Kryo will try to instantiate an ImmutableList first and then add elements to it.
I think this can be fixed in two ways:
1. Don't use immutable copies of the split offsets list. I am not sure what was the intention of making them immutable in the first place. Other metadata like metrics are still being provided in mutable collections.
2. Include custom Kryo serdes which support guava immutable collections. e.g. https://github.com/magro/kryo-serializers/

Option 2 seems a little overkill IMO for this specific case given that immutable collections don't seem to be necessary here, but would like to hear what others think.

### PR DESCRIPTION
**Pull request by rdblue**
Kryo can't handle guava ImmutableList and ImmutableMap. This commit avoids using those classes in DataFile. The trade-off is that DataFile instances that are copied more than once will not reuse the same immutable metrics maps.

Fixes #446

### CODE REVIEW

**Comment by jzhuge** [comment]
Thanks for the fix, let me try it.

**Comment by aokolnychyi** [comment]
Seems like there are some unused imports in `TestKryoSerialization`.

File: core/src/main/java/org/apache/iceberg/DataFiles.java [hunk 1]
```diff
@@ -334,4 +338,10 @@ public DataFile build() {
           keyMetadata, splitOffsets);
     }
   }
+
+  private static <E> List<E> copyList(List<E> toCopy) {
+    List<E> copy = Lists.newArrayListWithExpectedSize(toCopy.size());
+    copy.addAll(toCopy);
+    return copy;
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/DataFiles.java]: `GenericDataFile` also wraps the list into `Collections.unmodifiableList`. Does it make sense here?

File: spark/src/test/java/org/apache/iceberg/TestKryoSerialization.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        Assert.assertEquals("Should match the serialized record value counts",
+            dataFile.valueCounts(), result.valueCounts());
+        Assert.assertEquals("Should match the serialized record null value counts",
+            dataFile.nullValueCounts(), result.nullValueCounts());
+        Assert.assertEquals("Should match the serialized record lower bounds",
+            dataFile.lowerBounds(), result.lowerBounds());
+        Assert.assertEquals("Should match the serialized record upper bounds",
+            dataFile.upperBounds(), result.upperBounds());
+        Assert.assertEquals("Should match the serialized record key metadata",
+            dataFile.keyMetadata(), result.keyMetadata());
```
Comment (aokolnychyi) [inline spark/src/test/java/org/apache/iceberg/TestKryoSerialization.java]: Did we set something in `keyMetadata`? There is some custom serialization logic to handle this field.

Comment (rdblue) [reply spark/src/test/java/org/apache/iceberg/TestKryoSerialization.java]: No, this is empty.

Kryo can't handle ByteBuffers, but Spark's Kryo can. That's why this test uses Kryo provided by Spark. That's how this field is handled.

File: core/src/main/java/org/apache/iceberg/DataFiles.java (hunk 1 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/DataFiles.java]: I don't think it is necessary. In the long term, we should move these back to using ImmutableList. Right now, we just need to fix the serialization problem. There are other mutable instances being passed in from Metrics as well. We can fix all that later, but unblock the release now.

**Comment by rdblue** [comment]
Thanks @aokolnychyi!

**Pull request merged**